Sources/MongoSwift/BSON/Document+Collection.swift

```diff
@@ -0,0 +1,42 @@
+import Foundation
+
+/// An extension of `Document` to make it conform to the `Collection` protocol.
+/// This gives guarantees on non-destructive iteration, and offers an indexed
+/// ordering to the key value pairs in the document.
+extension Document: Collection {
+    /// Returns the start index of the Document.
+    public var startIndex: Int {
+        return 0
+    }
+
+    /// Returns the end index of the Document.
+    public var endIndex: Int {
+        precondition(self.count > 0)
+        return self.count
+    }
+
+    /// Returns the index after the given index for this Document.
+    public func index(after i: Int) -> Int {
+        // Index must be a valid one, meaning it must exist somewhere in self.keys.
+        _failEarlyRangeCheck(i, bounds: self.startIndex ... self.endIndex)
```

Comment:
where does this method come from?

Reply:
This is a method from `Collections.swift`. I originally had my own implementation until I realized that most likely Swift should and probably is doing the same exact thing I am in their default implementations. After some digging, I found that they do via uses of this function.

In retrospect, perhaps we shouldn't rely on this function, the leading underscore makes me a bit hesitant. I think at the time I was just happy to remove my own implementation for it, but it may be safer to rely on an implementation we have control over as opposed to this method which may or may not change.

Reply:
`_failEarlyRangeCheck` doesn't appear in any published documentation -- most search results are pointing to implementations (or code snippets) from standard library classes. I think this is worth recreating ourself as a private method.

Reply:
Yup, as I said, I think in retrospect its safer to rely on our own implementation.

Done!